### DISCUSSION TITLE
How to setup a workflow for downloading?

### DISCUSSION THREAD

**Discussion by ghost** [stashapp/stash#1873] [Q&A] [answered]
First of all let me say that I LOVE this app! Discovered it only yesterday and was amazed how smooth it was setting up!

There's one thing I can't wrap my head around though. I want to use yt-dlp to download new scenes on a regular basis (when I find something worth it).

Now I want the whole process to be as seamless as possible. Web-UI for yt-dlp where I copy the video url or even a Chrome extension to simply use a button click. yt-dlp then downloads the file and places it into Stash's data folder.

So far so good, but how do I get Stash to recognize the new file? Manually clicking scan can't be the solution. I'm fine with the fact that I have to go over the new files to clean up some metadata and so on.

On a side note - it would be nice to have the option so that newly discovered files are getting a specific tag like 'needs-review' so that I know which I have to check. Maybe worth a feature request?

Any suggestions on how to setup such a seamless workflow? I've seen that there are some plugins, but don't understand how to use them. Their documentation mostly consists of a one liner.

Hope someone can help me as I really love this project!

**Comment by philpw99** [comment]
There are a few ways to do what you want to do. For me, I will set up like this:
1. Use "4k Video Downloader" to paste the link and save it automatically in your collection folder, which is in the Stash "Settings->Configuration->Library". I am not familiar with yt-dlp so I will use the one I know.
2. Go to "Settings->Tasks->Scan" to let Stash recognize my newly downloaded videos. Or I can use my Stash_Helper to run scan at one click.
3. For new un-categorized videos, I will use a search filter: "Is Missing: Performers" to get the new ones. Then I will bookmark the link so next time I don't need to specify the filter again. If a scene has an unknown performer and I don't care who that is, I will just give a name "Nobody" to that. So next time that scene won't show up by the filter.
Right now I have 745 scenes and 622 movies organized by Stash this way. I am happy with what I did.

**Comment by ghost** [comment]
Hi phil, thanks for replying!
Unfortunately what you are discribing is not exactly what I'm looking for.

I do know that I can use one of the many frontends (be it web or desktop) to send a command to yt-dl which then downloads the video and saves it inside Stash's library folder. In fact that's what I'm already doing right now.

The workflow breaks when I have to manually start a Scan either from inside the Stash web interface or use a helper script/app to execute it. I want that part so be done automatically after yt-dl is finished downloading.

Is there a way to add a script to yt-dl that gets executed after every download? Maybe a downloader / web gui for it that supports something like that? (I'm not set on a gui for yt-dl yet, the only requirement is that it runs on Linux / Debian server and offers a web UI)

Using a filter to catch the new ones is a nice idea! Thanks, will look into that!

**Reply by 7dJx1qP** [reply]
Looks like yt-dl has an --exec option which can execute a script after a download is finished. https://unix.stackexchange.com/a/569088

**Reply by ghost** [reply]
Nice, thanks! So yt-dl could execute a script, thats good news. 
But does Stash provide a script or parameter to invoke scanning?

**Reply by ghost** [reply]
I still don't have a working solution, but wanted to post what I have anyway. Maybe it does help someone in the future.

To trigger a Stash scan you just invoke the following:
`curl -X POST -H "Content-Type: application/json" --data '{ "query": " mutation { metadataScan (input: { useFileMetadata: false} ) }" }' localhost:9998/graphql`

And in the youtube-dl config (or as parameter of the command itself):
`--exec CMD`

In total the command to download would look the following:
`./yt-dlp http://some-url.com/video --exec "curl -X POST -H "Content-Type: application/json" --data '{ "query": " mutation { metadataScan (input: { useFileMetadata: false} ) }" }' localhost:9998/graphql"`

But at the moment this throws a lot of errors when executing the command. I think I'll have to escape the command somehow but don't know how. Another problem could be that according to yt-dlp's documentation "--exec" appends the path to the downloaded file at the end.

Should I solve it, I'll update this. But the remaining issue seems to be with yt-dl and not stash.

**Reply by philpw99** [reply]
@Gishanx  You cannot Quote a quote. Notice after "--exec" there is a quote? After than the quote in -H "Content... lost the meaning.
Plus, it should be localhost:9999/graphql , not 9998
I think you can put the curl command in a cmd file, then --exec that cmd file.

**Reply by ghost** [reply]
> @Gishanx You cannot Quote a quote. Notice after "--exec" there is a quote? After than the quote in -H "Content... lost the meaning. Plus, it should be localhost:9999/graphql , not 9998 I think you can put the curl command in a cmd file, then --exec that cmd file.

**You are my hero!** And I don't mean that ironically!
Not because you explained escaping to me (was well aware of that if you read my last post).

But right at this moment I was sitting at a similar problem where simply putting the command in a file was the last piece to solve the puzzle. :)

Over the last few hours I've written my own webserver and bookmarklet to download with one click and invoke a Stash scan - all without having to manually intervene.

Will post how I did this tomorrow as I'm really tired now. Started the day working on my Arduino project and C# / WPF app and went down a rabbit hole learning python, flask, async webservers and more. What a day! :D

**Reply by philpw99** [reply]
What a busy day indeed ! For me knowing and doing that many kind of things and handle them is kinda curse and blessing at the same time. 😄

**Comment by ghost** [comment, marked as answer]
Here is my own solution for a more streamlined workflow: https://github.com/Gishanx/yt-dlp-api

It is far from complete and can be enhanced significantly but I'm happy with it for now.
Maybe there is someone out there who finds it useful too.

Oh and like I say in this repo - please bare with me if I made any major mistakes!
I'm very new to all of this - I began learning about Python, web servers and so much more only yesterday.
And my experience with Linux is also very limited.

So please if there are some major flaws with this please point them out so I can learn! Thanks!

**Reply by xx790** [reply]
The user was deleted. I wonder whether this has anything to do with yt-dlp being "problematic".
I thought about implementing a similar thing and I'm in doubt whether I can make it public.
